File: recoverBinarySearchTree.py
import collections
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
  # @param root, a tree node
  # @return a tree node
  def recoverTree(self, root):
    n = root
    p = None
    large = None
    while n:
      if n.left:
        m = n.left
        while m.right:
          m = m.right

        l = n.left
        n.left = None
        m.right = n
        n = l
        root = n

      else:
        print n.val
        n = n.right

    return root


# test code 
t1 = TreeNode(4)
t2 = TreeNode(2)
t3 = TreeNode(5)
t4 = TreeNode(1)
t5 = TreeNode(3)
t6 = TreeNode(6)
t7 = TreeNode(7)

t1.left = t2
t1.right = t3
t2.left = t4
t2.right = t5
t3.left = t6
t3.right = t7

s = Solution()
root = s.recoverTree(t1)
print root.val

print t1.left
print t1.right.val
print t2.left
print t2.right.val
print t3.left
print t3.right.val
print t4.left
print t4.right.val
print t5.left
print t5.right.val
print t6.left
print t6.right.val
print t7.left
print t7.right

print '====='

q = collections.deque()
q.append(t1)
while q:
  n = q.popleft()
  print n.val
  if n.left:
    print n.val,'.left', n.left.val
    q.append(n.left)

  if n.right:
    print n.val, '.right', n.right.val
    q.append(n.right)

File: recover_binary_search_tree.rb
# Definition for a binary tree node.
class TreeNode
    attr_accessor :val, :left, :right
    def initialize(val)
        @val = val
        @left, @right = nil, nil
    end
end

# @param {TreeNode} root
# @return {Void} Do not return anything, modify root in-place instead.
def recover_tree(root)
  return if root.nil?

  stack = []

  prev = TreeNode.new(-Float::INFINITY)
  n = root
  damn_node = nil

  swap_nodes = []

  while !n.nil? || !stack.empty? do
    if n.nil?
      n = stack.pop

      if !prev.nil?
        if n.val < prev.val
          if swap_nodes.empty?
            swap_nodes << prev
            damn_node = n
          else
            swap_nodes << n
          end
        end
      end

      prev = n

      n = n.right
    else
      while !n.nil? do
        stack << n
        n = n.left
      end
    end
  end

  if swap_nodes.length < 2
    if !damn_node.nil?
      swap_nodes << damn_node
    elsif swap_nodes[0] == root
      if root.left && root.left.val > root.val
        swap_nodes << root.left
      elsif root.right && root.right.val < root.val
        swap_nodes << root.right
      end
    else
      swap_nodes << root
    end
  end


  tmp = swap_nodes[0].val
  swap_nodes[0].val = swap_nodes[1].val
  swap_nodes[1].val = tmp
  nil
end
